pkg/util/timeout.go

```diff
@@ -0,0 +1,37 @@
+package util
+
+import (
+	"fmt"
+	"time"
+)
+
+// TimeoutError is error returned after timeout occured.
+type TimeoutError struct {
+	after time.Duration
+}
+
+// Error implements the Go error interface.
+func (t *TimeoutError) Error() string {
+	return fmt.Sprintf("calling the function timeout after %v", t.after)
+}
+
+// TimeoutAfter executes the provide function and return the TimeoutError in
+// case when the execution time of the provided function is bigger than provided
+// time duration.
+func TimeoutAfter(t time.Duration, fn func() error) error {
+	c := make(chan error, 1)
+	defer close(c)
```

Comment:
can you provide an example where this will panic?
On Mar 17, 2016 3:35 PM, "Michal Fojtik" <mi@mifo.sk> wrote:

> i dont think it will panic... that will show up in unot test, right?
> On Mar 17, 2016 3:18 PM, "Rodolfo Carvalho" <notifications@github.com>
> wrote:
>
>> In pkg/util/timeout.go
>> <https://github.com/openshift/source-to-image/pull/429#discussion_r56510436>
>> :
>>
>> > +// TimeoutError is error returned after timeout occured.
>> > +type TimeoutError struct {
>> > +	after time.Duration
>> > +}
>> > +
>> > +// Error implements the Go error interface.
>> > +func (t *TimeoutError) Error() string {
>> > +	return fmt.Sprintf("calling the function timeout after %v", t.after)
>> > +}
>> > +
>> > +// TimeoutAfter executes the provide function and return the TimeoutError in
>> > +// case when the execution time of the provided function is bigger than provided
>> > +// time duration.
>> > +func TimeoutAfter(t time.Duration, fn func() error) error {
>> > +	c := make(chan error, 1)
>> > +	defer close(c)
>>
>> The buffer in the channel is the right pattern to avoid leaking a
>> goroutine, well done :)
>>
>> There's still a bug here. When we time out, the goroutine started in the
>> next line will panic sending on a closed channel.
>>
>> It should be:
>>
>> c := make(chan error, 1)go func() {
>>   defer close(c)
>>   c <- fn()
>> }()select {
>>
>> That ensures the channel is only closed after we send.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/source-to-image/pull/429/files/edbb438d2caca4b8e3721d6c29a65ac289e5d1e3#r56510436>
>>
>